File: src/test/java/com/github/ccguyka/showupdates/consumer/ProjectUpdatesMergerTest.java
package com.github.ccguyka.showupdates.consumer;

import static com.google.common.collect.Lists.newArrayList;
import static org.assertj.core.api.Assertions.assertThat;
import org.junit.Test;

import com.github.ccguyka.showupdates.objects.ArtifactUpdate;
import com.github.ccguyka.showupdates.objects.DependencyUpdates;
import com.github.ccguyka.showupdates.objects.ProjectUpdates;

public class ProjectUpdatesMergerTest {

    private ProjectUpdatesMerger merger = new ProjectUpdatesMerger();

    @Test
    public void mergeParent() {
        ProjectUpdates first = ProjectUpdates.builder()
                .withParent(new DependencyUpdates(newArrayList(new ArtifactUpdate("par1", "1", newArrayList("2")),
                        new ArtifactUpdate("par2", "2", newArrayList("11")))))
                .build();
        ProjectUpdates second = ProjectUpdates.builder()
                .withParent(new DependencyUpdates(newArrayList(new ArtifactUpdate("par1", "1", newArrayList("2")),
                        new ArtifactUpdate("par3", "5", newArrayList("6")))))
                .build();

        ProjectUpdates result = merger.merge(first, second);

        ProjectUpdates expected = ProjectUpdates.builder()
                .withParent(new DependencyUpdates(newArrayList(new ArtifactUpdate("par1", "1", newArrayList("2")),
                        new ArtifactUpdate("par2", "2", newArrayList("11")),
                        new ArtifactUpdate("par3", "5", newArrayList("6")))))
            .build();
        assertThat(result.getParent().getArtifacts())
            .containsExactlyInAnyOrderElementsOf(expected.getParent().getArtifacts());
    }

    @Test
    public void mergeDependency() {
        ProjectUpdates first = ProjectUpdates.builder()
                .withDependency(new DependencyUpdates(newArrayList(new ArtifactUpdate("dep1", "1", newArrayList("2")),
                        new ArtifactUpdate("dep2", "2", newArrayList("11")))))
                .build();
        ProjectUpdates second = ProjectUpdates.builder()
                .withDependency(new DependencyUpdates(newArrayList(new ArtifactUpdate("dep1", "1", newArrayList("2")),
                        new ArtifactUpdate("dep3", "5", newArrayList("6")))))
                .build();

        ProjectUpdates result = merger.merge(first, second);

        ProjectUpdates expected = ProjectUpdates.builder()
                .withDependency(new DependencyUpdates(newArrayList(new ArtifactUpdate("dep1", "1", newArrayList("2")),
                        new ArtifactUpdate("dep2", "2", newArrayList("11")),
                        new ArtifactUpdate("dep3", "5", newArrayList("6")))))
            .build();
        assertThat(result.getDependency().getArtifacts())
            .containsExactlyInAnyOrderElementsOf(expected.getDependency().getArtifacts());
    }

    @Test
    public void mergeDependencyManagement() {
        ProjectUpdates first = ProjectUpdates.builder()
                .withDependencyManagement(new DependencyUpdates(newArrayList(new ArtifactUpdate("depMgnt1", "1", newArrayList("2")),
                        new ArtifactUpdate("depMgnt2", "2", newArrayList("11")))))
                .build();
        ProjectUpdates second = ProjectUpdates.builder()
                .withDependencyManagement(new DependencyUpdates(newArrayList(new ArtifactUpdate("depMgnt1", "1", newArrayList("2")),
                        new ArtifactUpdate("depMgnt3", "5", newArrayList("6")))))
                .build();

        ProjectUpdates result = merger.merge(first, second);

        ProjectUpdates expected = ProjectUpdates.builder()
                .withDependencyManagement(new DependencyUpdates(newArrayList(new ArtifactUpdate("depMgnt1", "1", newArrayList("2")),
                        new ArtifactUpdate("depMgnt2", "2", newArrayList("11")),
                        new ArtifactUpdate("depMgnt3", "5", newArrayList("6")))))
            .build();
        assertThat(result.getDependencyManagement().getArtifacts())
            .containsExactlyInAnyOrderElementsOf(expected.getDependencyManagement().getArtifacts());
    }

    @Test
    public void mergePlugin() {
        ProjectUpdates first = ProjectUpdates.builder()
                .withPlugin(new DependencyUpdates(newArrayList(new ArtifactUpdate("plugin1", "1", newArrayList("2")),
                        new ArtifactUpdate("plugin2", "2", newArrayList("11")))))
                .build();
        ProjectUpdates second = ProjectUpdates.builder()
                .withPlugin(new DependencyUpdates(newArrayList(new ArtifactUpdate("plugin1", "1", newArrayList("2")),
                        new ArtifactUpdate("plugin3", "5", newArrayList("6")))))
                .build();

        ProjectUpdates result = merger.merge(first, second);

        ProjectUpdates expected = ProjectUpdates.builder()
                .withPlugin(new DependencyUpdates(newArrayList(new ArtifactUpdate("plugin1", "1", newArrayList("2")),
                        new ArtifactUpdate("plugin2", "2", newArrayList("11")),
                        new ArtifactUpdate("plugin3", "5", newArrayList("6")))))
            .build();
        assertThat(result.getPlugin().getArtifacts())
            .containsExactlyInAnyOrderElementsOf(expected.getPlugin().getArtifacts());
    }
}

File: src/test/kotlin/com/github/ccguyka/showupdates/consumer/ProjectUpdatesMergerTest.kt
package com.github.ccguyka.showupdates.consumer

import com.github.ccguyka.showupdates.objects.ArtifactUpdate
import com.github.ccguyka.showupdates.objects.DependencyUpdates
import com.github.ccguyka.showupdates.objects.ProjectUpdates.Companion.builder
import com.google.common.collect.Lists
import org.assertj.core.api.Assertions
import org.junit.Test

class ProjectUpdatesMergerTest {

    private val merger = ProjectUpdatesMerger()

    @Test
    fun mergeParent() {
        val first = builder()
                .withParent(DependencyUpdates(Lists.newArrayList(ArtifactUpdate("par1", "1", Lists.newArrayList("2")),
                        ArtifactUpdate("par2", "2", Lists.newArrayList("11")))))
                .build()
        val second = builder()
                .withParent(DependencyUpdates(Lists.newArrayList(ArtifactUpdate("par1", "1", Lists.newArrayList("2")),
                        ArtifactUpdate("par3", "5", Lists.newArrayList("6")))))
                .build()
        val result = merger.merge(first, second)
        val expected = builder()
                .withParent(DependencyUpdates(Lists.newArrayList(ArtifactUpdate("par1", "1", Lists.newArrayList("2")),
                        ArtifactUpdate("par2", "2", Lists.newArrayList("11")),
                        ArtifactUpdate("par3", "5", Lists.newArrayList("6")))))
                .build()
        Assertions.assertThat(result.parent.artifacts)
                .containsExactlyInAnyOrderElementsOf(expected.parent.artifacts)
    }

    @Test
    fun mergeDependency() {
        val first = builder()
                .withDependency(DependencyUpdates(Lists.newArrayList(ArtifactUpdate("dep1", "1", Lists.newArrayList("2")),
                        ArtifactUpdate("dep2", "2", Lists.newArrayList("11")))))
                .build()
        val second = builder()
                .withDependency(DependencyUpdates(Lists.newArrayList(ArtifactUpdate("dep1", "1", Lists.newArrayList("2")),
                        ArtifactUpdate("dep3", "5", Lists.newArrayList("6")))))
                .build()
        val result = merger.merge(first, second)
        val expected = builder()
                .withDependency(DependencyUpdates(Lists.newArrayList(ArtifactUpdate("dep1", "1", Lists.newArrayList("2")),
                        ArtifactUpdate("dep2", "2", Lists.newArrayList("11")),
                        ArtifactUpdate("dep3", "5", Lists.newArrayList("6")))))
                .build()
        Assertions.assertThat(result.dependency.artifacts)
                .containsExactlyInAnyOrderElementsOf(expected.dependency.artifacts)
    }

    @Test
    fun mergeDependencyManagement() {
        val first = builder()
                .withDependencyManagement(DependencyUpdates(Lists.newArrayList(ArtifactUpdate("depMgnt1", "1", Lists.newArrayList("2")),
                        ArtifactUpdate("depMgnt2", "2", Lists.newArrayList("11")))))
                .build()
        val second = builder()
                .withDependencyManagement(DependencyUpdates(Lists.newArrayList(ArtifactUpdate("depMgnt1", "1", Lists.newArrayList("2")),
                        ArtifactUpdate("depMgnt3", "5", Lists.newArrayList("6")))))
                .build()
        val result = merger.merge(first, second)
        val expected = builder()
                .withDependencyManagement(DependencyUpdates(Lists.newArrayList(ArtifactUpdate("depMgnt1", "1", Lists.newArrayList("2")),
                        ArtifactUpdate("depMgnt2", "2", Lists.newArrayList("11")),
                        ArtifactUpdate("depMgnt3", "5", Lists.newArrayList("6")))))
                .build()
        Assertions.assertThat(result.dependencyManagement.artifacts)
                .containsExactlyInAnyOrderElementsOf(expected.dependencyManagement.artifacts)
    }

    @Test
    fun mergePlugin() {
        val first = builder()
                .withPlugin(DependencyUpdates(Lists.newArrayList(ArtifactUpdate("plugin1", "1", Lists.newArrayList("2")),
                        ArtifactUpdate("plugin2", "2", Lists.newArrayList("11")))))
                .build()
        val second = builder()
                .withPlugin(DependencyUpdates(Lists.newArrayList(ArtifactUpdate("plugin1", "1", Lists.newArrayList("2")),
                        ArtifactUpdate("plugin3", "5", Lists.newArrayList("6")))))
                .build()
        val result = merger.merge(first, second)
        val expected = builder()
                .withPlugin(DependencyUpdates(Lists.newArrayList(ArtifactUpdate("plugin1", "1", Lists.newArrayList("2")),
                        ArtifactUpdate("plugin2", "2", Lists.newArrayList("11")),
                        ArtifactUpdate("plugin3", "5", Lists.newArrayList("6")))))
                .build()
        Assertions.assertThat(result.plugin.artifacts)
                .containsExactlyInAnyOrderElementsOf(expected.plugin.artifacts)
    }
}
